### PR TITLE
feat: DSpark speculative decoding draft model training for RL rollout acceleration

### PR DESCRIPTION
**Pull request by CalvinXKY**
## Summary

Integrates DSpark speculative decoding into vime's RL training pipeline to accelerate rollout generation. DSpark trains a lightweight draft model (Markov heads + attention layers) alongside the policy model, enabling speculative decoding during vLLM rollout with ~2.5x throughput improvement.

### Key Changes

**New module: `vime/backends/megatron_utils/dspark/`**
- Draft model architecture: Markov heads, attention layers, hidden state capture
- Training losses: CE + L1 (chunked logsumexp for memory efficiency) + confidence loss
- Weight export to vLLM for speculative decoding during rollout
- TP>1 support with vocab padding/unpadding (all-gather, zero-pad, strip)

**Actor integration (`actor.py`)**
- Draft weight cache/restore lifecycle around sleep/wake_up to prevent GPU OOM
- `synchronize()+empty_cache()` before `torch_memory_saver.pause()` to prevent memory fragmentation crash
- Unified try/except for pause/resume across all DSpark configs

**Weight sync (`update_weight/common.py`, `update_weight_from_distributed.py`, `update_weight_from_tensor.py`)**
- vLLM 0.26.0/0.27.2 compatibility shims: `NCCLTrainerAdapter` with `StatelessProcessGroup` + `trainer_send_weights`
- `create_nccl_trainer` tries vLLM 0.27.2 API first, falls back to 0.26.0 adapter
- `send_custom_weights` sends metadata via Ray before NCCL broadcast
- DSpark draft weight sync via file-based transfer (`load_draft_weights_from_file` through `/collective_rpc` HTTP endpoint)
- No-op `cache_dspark_draft_weights` for non-colocate mode

**vLLM engine (`vllm_engine.py`)**
- Engine router retry + extended timeouts for DSpark weight sync
- spec_accept_rate extraction from `/inference/v1/generate` endpoint
- `load_draft_weights_from_file` method for DSpark draft model weight loading

**CLI arguments (`arguments.py`)**
- 14 new `--dspark-*` flags for draft model config (layers, block_size, num_speculative_tokens, loss weights, etc.)

## Test plan
- [x] Unit tests: `tests/test_dspark_unit.py` (model build, forward, loss)
- [x] Phase 2 integration test: `tests/test_dspark_phase2.py` (end-to-end training loop)
- [x] TP=1: Qwen3-4B 200-step GRPO, DSpark ON vs OFF, AIME 2024 eval
- [x] TP=4: Qwen3-4B DSpark training verified (24.87% spec_accept_rate)
- [x] Colocate 20-step test (vLLM 0.26.0, origin/main rebase): spec_accept_rate 32-37%
- [x] Non-colocate 20-step test: spec_accept_rate 33-37%
- [x] Colocate 100-step long run: spec_accept_rate 32-40%, ce_loss 3.27→1.68 (draft model learned)
- [x] vLLM 0.27.2 verification: weight sync (IPC packed + draft file transfer) verified, spec_accept_rate 32-36%, AIME 58.75%
- [ ] CI pipeline validation
- [ ] Multi-model validation (Qwen3-8B, etc.)

## Notes
- Requires vLLM 0.26.0+ (compatible with both 0.26.0 and 0.27.2 via try/except shims)
- vLLM patch needed: add `cos_sin_cache` to `SKIP_TENSORS` in `reload/meta.py` (DSpark draft export has no buffers)
- Pretrained draft model initialization recommended (random init shows ce_loss plateau at ~12)
- vLLM 0.27.2 image (39edbd2ae25a) has TE 2.16.1 kernels incompatible with A800 (SM 8.0) — use vLLM 0.26.0 image for A800 training

🤖 Generated with CodeAgent

### CODE REVIEW

**Comment by CalvinXKY** [comment]
image

**Comment by CalvinXKY** [comment]
## 4B training （colocate）


```
#!/bin/bash
# DSpark 4B 100-step colocate long run (rebased on origin/main)
# Mode: colocate- training and rollout share same GPUs
# Goal: Verify DSpark stability over 100 steps without eval

set -ex

# Cleanup
pkill -9 -f '[v]llm serve|VLL[M]::' || true
sleep 2
ray stop --force || true
pkill -9 ray || true
pkill -9 python || true
sleep 2

export PYTHONUNBUFFERED=1

NUM_GPUS=8

# Model config (Qwen3-4B)
MODEL_ARGS=(
   --swiglu
   --num-layers 36
   --hidden-size 2560
   --ffn-hidden-size 9728
   --num-attention-heads 32
   --group-query-attention
   --num-query-groups 8
   --use-rotary-position-embeddings
   --disable-bias-linear
   --normalization "RMSNorm"
   --norm-epsilon 1e-6
   --rotary-base 1000000
   --vocab-size 151936
   --kv-channels 128
   --qk-layernorm
)

CKPT_ARGS=(
   --hf-checkpoint /data/nfs_87/xky/models/Qwen3-4B
   --ref-load /data/nfs_87/xky/models/Qwen3-4B_torch_dist
   --save /data/nfs_87/xky/dspark_upgrade/save_4b_colocate_100step
   --save-interval 100
)

ROLLOUT_ARGS=(
   --prompt-data /data/nfs_87/datasets/dapo-math-17k/dapo-math-17k.jsonl
   --input-key prompt
   --label-key label
   --apply-chat-template
   --rollout-shuffle
   --rm-type deepscaler
   --num-rollout 100
   --rollout-batch-size 32
   --n-samples-per-prompt 8
   --rollout-max-response-len 4096
   --rollout-temperature 1
   --global-batch-size 256
   --balance-data
)

PERF_ARGS=(
   --tensor-model-parallel-size 1
   --pipeline-model-parallel-size 1
   --context-parallel-size 1
   --expert-model-parallel-size 1
   --expert-tensor-parallel-size 1
   --recompute-granularity full
   --recompute-method uniform
   --recompute-num-layers 1
   --use-dynamic-batch-size
   --max-tokens-per-gpu 9216
)

GRPO_ARGS=(
   --advantage-estimator grpo
   --use-kl-loss
   --kl-loss-coef 0.00
   --kl-loss-type low_var_kl
   --entropy-coef 0.00
   --eps-clip 0.2
   --eps-clip-high 0.28
)

OPTIMIZER_ARGS=(
   --optimizer adam
   --lr 1e-6
   --lr-decay-style constant
   --weight-decay 0.1
   --adam-beta1 0.9
   --adam-beta2 0.98
)

VLLM_ARGS=(
   --rollout-num-gpus-per-engine 1
   --vllm-gpu-memory-utilization 0.5
   --vllm-enforce-eager
   --vllm-speculative-config '{"method":"dspark","model":"/data/nfs_87/xky/models/Qwen3-4B-dspark-pretrained","num_speculative_tokens":7}'
)

MISC_ARGS=(
   --attention-dropout 0.0
   --hidden-dropout 0.0
   --accumulate-allreduce-grads-in-fp32
   --attention-softmax-in-fp32
   --attention-backend flash
)

DSPARK_ARGS=(
   --dspark-enabled
   --dspark-block-size 7
   --dspark-num-draft-layers 5
   --dspark-target-layer-ids 1,9,17,25,33
   --dspark-markov-rank 256
   --dspark-markov-head-type vanilla
   --dspark-num-anchors 512
   --dspark-mask-token-id 151669
   --dspark-ce-loss-alpha 0.5
   --dspark-l1-loss-alpha 0.5
   --dspark-confidence-head-alpha 0.1
   --dspark-loss-decay-gamma 4.0
   --dspark-draft-loss-weight 1.0
   --dspark-intermediate-size 9728
   --dspark-pretrained-model /data/nfs_87/xky/models/Qwen3-4B-dspark-pretrained/model.safetensors
)

# Launch ray
ray start --head --node-ip-address 127.0.0.1 --num-gpus ${NUM_GPUS} --disable-usage-stats --dashboard-host=0.0.0.0 --dashboard-port=8265

VIME_DIR="/data/nfs_87/xky/dspark_upgrade/vime"

RUNTIME_ENV_JSON="{
  \"working_dir\": \"${VIME_DIR}\",
  \"env_vars\": {
    \"PYTHONPATH\": \"/root/Megatron-LM/\",
    \"CUDA_DEVICE_MAX_CONNECTIONS\": \"1\",
    \"NCCL_NVLS_ENABLE\": \"1\"
  }
}"

ray job submit --address="http://127.0.0.1:8265" \
   --runtime-env-json="${RUNTIME_ENV_JSON}" \
   -- python3 train.py \
   --actor-num-nodes 1 \
   --actor-num-gpus-per-node ${NUM_GPUS} \
   --colocate \
   ${MODEL_ARGS[@]} \
   ${CKPT_ARGS[@]} \
   ${ROLLOUT_ARGS[@]} \
   ${OPTIMIZER_ARGS[@]} \
   ${GRPO_ARGS[@]} \
   ${PERF_ARGS[@]} \
   ${VLLM_ARGS[@]} \
   ${MISC_ARGS[@]} \
   ${DSPARK_ARGS[@]}

```

## result

image

image

image

image

**Comment by CalvinXKY** [comment]
## non-colocate mode sh example

main commit id :  8ca82e0

```
#!/bin/bash
# DSpark 4B 20-step non-colocate test (rebased on origin/main)
# Mode: non-colocate - training GPUs and rollout GPUs are separate
# Goal: Verify DSpark works in non-colocate mode for 20 steps
# Layout: 4 GPUs for actor training + 4 GPUs for rollout inference = 8 total

set -ex

# Cleanup
pkill -9 -f '[v]llm serve|VLL[M]::' || true
sleep 2
ray stop --force || true
pkill -9 ray || true
pkill -9 python || true
sleep 2

export PYTHONUNBUFFERED=1

ACTOR_NUM_GPUS=4
ROLLOUT_NUM_GPUS=4

# Model config (Qwen3-4B)
MODEL_ARGS=(
   --swiglu
   --num-layers 36
   --hidden-size 2560
   --ffn-hidden-size 9728
   --num-attention-heads 32
   --group-query-attention
   --num-query-groups 8
   --use-rotary-position-embeddings
   --disable-bias-linear
   --normalization "RMSNorm"
   --norm-epsilon 1e-6
   --rotary-base 1000000
   --vocab-size 151936
   --kv-channels 128
   --qk-layernorm
)

CKPT_ARGS=(
   --hf-checkpoint /data/nfs_87/xky/models/Qwen3-4B
   --ref-load /data/nfs_87/xky/models/Qwen3-4B_torch_dist
   --save /data/nfs_87/xky/dspark_upgrade/save_4b_noncolocate_20step
   --save-interval 100
)

ROLLOUT_ARGS=(
   --prompt-data /data/nfs_87/datasets/dapo-math-17k/dapo-math-17k.jsonl
   --input-key prompt
   --label-key label
   --apply-chat-template
   --rollout-shuffle
   --rm-type deepscaler
   --num-rollout 20
   --rollout-batch-size 32
   --n-samples-per-prompt 8
   --rollout-max-response-len 4096
   --rollout-temperature 1
   --global-batch-size 256
   --balance-data
)

PERF_ARGS=(
   --tensor-model-parallel-size 1
   --pipeline-model-parallel-size 1
   --context-parallel-size 1
   --expert-model-parallel-size 1
   --expert-tensor-parallel-size 1
   --recompute-granularity full
   --recompute-method uniform
   --recompute-num-layers 1
   --use-dynamic-batch-size
   --max-tokens-per-gpu 9216
)

GRPO_ARGS=(
   --advantage-estimator grpo
   --use-kl-loss
   --kl-loss-coef 0.00
   --kl-loss-type low_var_kl
   --entropy-coef 0.00
   --eps-clip 0.2
   --eps-clip-high 0.28
)

OPTIMIZER_ARGS=(
   --optimizer adam
   --lr 1e-6
   --lr-decay-style constant
   --weight-decay 0.1
   --adam-beta1 0.9
   --adam-beta2 0.98
)

VLLM_ARGS=(
   --rollout-num-gpus-per-engine 1
   --rollout-num-gpus ${ROLLOUT_NUM_GPUS}
   --vllm-gpu-memory-utilization 0.85
   --vllm-enforce-eager
   --vllm-speculative-config '{"method":"dspark","model":"/data/nfs_87/xky/models/Qwen3-4B-dspark-pretrained","num_speculative_tokens":7}'
)

MISC_ARGS=(
   --attention-dropout 0.0
   --hidden-dropout 0.0
   --accumulate-allreduce-grads-in-fp32
   --attention-softmax-in-fp32
   --attention-backend flash
)

DSPARK_ARGS=(
   --dspark-enabled
   --dspark-block-size 7
   --dspark-num-draft-layers 5
   --dspark-target-layer-ids 1,9,17,25,33
   --dspark-markov-rank 256
   --dspark-markov-head-type vanilla
   --dspark-num-anchors 512
   --dspark-mask-token-id 151669
   --dspark-ce-loss-alpha 0.5
   --dspark-l1-loss-alpha 0.5
   --dspark-confidence-head-alpha 0.1
   --dspark-loss-decay-gamma 4.0
   --dspark-draft-loss-weight 1.0
   --dspark-intermediate-size 9728
   --dspark-pretrained-model /data/nfs_87/xky/models/Qwen3-4B-dspark-pretrained/model.safetensors
)

# Launch ray with total GPUs = actor + rollout
TOTAL_GPUS=$((ACTOR_NUM_GPUS + ROLLOUT_NUM_GPUS))
ray start --head --node-ip-address 127.0.0.1 --num-gpus ${TOTAL_GPUS} --disable-usage-stats --dashboard-host=0.0.0.0 --dashboard-port=8265

VIME_DIR="/data/nfs_87/xky/dspark_upgrade/vime"

RUNTIME_ENV_JSON="{
  \"working_dir\": \"${VIME_DIR}\",
  \"env_vars\": {
    \"PYTHONPATH\": \"/root/Megatron-LM/\",
    \"CUDA_DEVICE_MAX_CONNECTIONS\": \"1\",
    \"NCCL_NVLS_ENABLE\": \"1\"
  }
}"

ray job submit --address="http://127.0.0.1:8265" \
   --runtime-env-json="${RUNTIME_ENV_JSON}" \
   -- python3 train.py \
   --actor-num-nodes 1 \
   --actor-num-gpus-per-node ${ACTOR_NUM_GPUS} \
   ${MODEL_ARGS[@]} \
   ${CKPT_ARGS[@]} \
   ${ROLLOUT_ARGS[@]} \
   ${OPTIMIZER_ARGS[@]} \
   ${GRPO_ARGS[@]} \
   ${PERF_ARGS[@]} \
   ${VLLM_ARGS[@]} \
   ${MISC_ARGS[@]} \
   ${DSPARK_ARGS[@]}

```

**Review by Meihan-chen — DISMISSED** [review]

**Comment by CalvinXKY** [comment]
## Tests with Commit e4e51ac 

Loss doesn’t look normal:

image

Figure out where the problem lies, next~~

**Review by Meihan-chen — APPROVED** [review]

**Comment by CalvinXKY** [comment]
> ## Tests with Commit [e4e51ac](https://github.com/vllm-project/vime/commit/e4e51acb1858965894a16bc460fe9236d8e66e6c)
> Loss doesn’t look normal:
> 
> image
> Figure out where the problem lies, next~~

## Need fix: after optimizer.step(), copy the draft parameters from the DDP buffer back to param.data.

## Result

image
